lib/Conversion/SeqToSV/SeqToSV.cpp

```diff
@@ -448,26 +423,58 @@ void SeqToSVPass::runOnOperation() {
   // Lower memories and registers in modules in parallel.
   bool needsRegRandomization = false;
   bool needsMemRandomization = false;
+  DenseMap<HWModuleOp, std::pair<bool, bool>> moduleFragments;
+  llvm::sys::SmartMutex<true> fragmentsMutex;
   mlir::parallelForEach(&getContext(), modules, [&](HWModuleOp module) {
     SeqToSVTypeConverter typeConverter;
     FirRegLowering regLowering(typeConverter, module, disableRegRandomization,
                                emitSeparateAlwaysBlocks);
     regLowering.lower();
     if (regLowering.needsRegRandomization()) {
-      if (!disableRegRandomization)
-        addFragments(module, regFragments);
+      if (!disableRegRandomization) {
+        llvm::sys::SmartScopedLock<true> lock(fragmentsMutex);
+        moduleFragments[module].first = true;
+      }
       needsRegRandomization = true;
     }
     numSubaccessRestored += regLowering.numSubaccessRestored;
 
     if (auto *it = memsByModule.find(module); it != memsByModule.end()) {
       memLowering.lowerMemoriesInModule(module, it->second);
-      if (!disableMemRandomization)
-        addFragments(module, memFragments);
+      if (!disableMemRandomization) {
+        llvm::sys::SmartScopedLock<true> lock(fragmentsMutex);
+        moduleFragments[module].second = true;
+      }
       needsMemRandomization = true;
     }
   });
 
+  auto randomInitFragmentName =
+      FlatSymbolRefAttr::get(context, "RANDOM_INIT_FRAGMENT");
+  auto randomInitRegFragmentName =
+      FlatSymbolRefAttr::get(context, "RANDOM_INIT_REG_FRAGMENT");
+  auto randomInitMemFragmentName =
+      FlatSymbolRefAttr::get(context, "RANDOM_INIT_MEM_FRAGMENT");
```

Comment:
Really nitty, however, these identifiers are all very similar and it makes it slightly difficult to check their usage later. Shorter names would make it more obvious that things like the following are correct without having to scan into the middle of the identifier:

``` c++
if (needsReg)
  fragmentAttrs.push_back(regFragment);
```